active/0000-command-line-completion.md

```diff
@@ -0,0 +1,106 @@
+- ember-cli Issue: (leave this empty)
+
+# Summary
+
+Adds command line completion *(tab completion)* to ember-cli that fills in partially typed commands by the user and suggests available sub commands or options. *(from now on i will refer to "strings in the command line" as "cli-commands" and "generate" and "new" as unprefixed "commands")*
+
+![](http://zippy.gfycat.com/PrestigiousLoneHalicore.gif)
+
+# Motivation
+
+With all the already existing commands and especially all blueprints, plus the fact that any addon can add even more blueprints to your tool kit, users can get overwhelmed. Currently, when you want to execute a specific task and you don't quite know the correct cli-command you have to invoke `ember help` which is noisy and slow *( especially when you just want to know the spelling of a specific thing)*. This feature will enable the user to choose from all existing cli-commands by pressing __[tab]__ or just to let ember-cli fill partially typed cli-commands for speed.
+
+# Detailed design
+
+The two main components of that feature are a __completion function__ that is responsible for the actual tab-completion, and a __generation function__ that will write the cli-command hierarchy together with metadata into a JSON file for fast processing.
+
+## completion function
+
+To enable this feature, a __completion function__ will run at the main entry point of the process *(making use of  [omelette](https://github.com/f/omelettev) for shell completion)*. On every __[tab]__ it will parse the command line and either completes a partially typed, unambiguous cli-command, or suggests possible cli-commands for the current context.
+
+The user interface will work as you would expect it from a shell completion:
+
+- it suggests all commands if none are typed yet
+
+  ```
+    $ ember <tab>
+    > addon     destroy   help      install   serve     version
+      build     generate  init      new       test
+  ```
+- it completes partially typed commands
+
+  ```
+    $ ember gen<tab>
+    $ ember generate
+  ```
+- it completes to suggests commands based on user input (note how it __should__ understand aliases)
+
+  ```
+    $ ember g ad<tab>
+    > adapter       adapter-test  addon
+  ```
+- it, by default, will not suggest options
+
+  ```
+    $ ember g resource <tab>
+  ```
+- it will suggest options on demand (note how it __should__ know when an option needs a value)
+
+  ```
+    $ ember g resource post --<tab>
+    --dry-run         --in-repo-addon=  --verbose
+    --dummy           --pod
+  ```
+
+## generation function
+
+For a good user experience we don't want to figure out those suggestions on runtime or the completion feature would not be substantially faster then the `ember help` command. So there will be a __generation function__ that generates a JSON file once after ember-cli is installed and then during every `ember install some-addon` command to ensure that blueprints added by new addons are recognized aswell.
```

Comment:
The cache key should be described in more detail, `ember install` is only run on the installers machine now on there co-workers machines.

It may be interesting to compute a runtime cache key that represents all add-ons and there versions?

It may also be ok to take different approaches in insuring this is quick enough that doesn't require caching at all.